flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClientTest.java

```diff
@@ -454,6 +451,22 @@ void testWatchPodsAndDoCallback() throws Exception {
         assertThat(podModifiedAction.get()).isEqualTo(Action.MODIFIED);
     }
 
+    @Test
+    void testWatchPodsAndDoCallbackFail() throws Exception {
+        mockPodEventWithLabelsFail(
+                NAMESPACE, TASKMANAGER_POD_NAME, KUBERNETES_ZERO_RESOURCE_VERSION, TESTING_LABELS);
+        TestingWatchCallbackHandler<KubernetesPod> watchCallbackHandler =
+                TestingWatchCallbackHandler.<KubernetesPod>builder().build();
+        Long watchStartTime = System.currentTimeMillis();
+        assertThatThrownBy(
+                () ->
+                        this.flinkKubeClient.watchPodsAndDoCallback(
+                                TESTING_LABELS, watchCallbackHandler));
+        Long watchEndTime = System.currentTimeMillis();
+        Long watchDuration = watchEndTime - watchStartTime;
+        assertThat(watchDuration > 100);
```

Comment:
Relying on the duration to decide whether the retry has happened can be unstable. I'd suggest to mock the server in a way that replies failure messages for certain times following by a success message.